internals/daemon/daemon.go

```diff
@@ -717,21 +717,55 @@ func (d *Daemon) doReboot(sigCh chan<- os.Signal, waitTimeout time.Duration) err
 	return fmt.Errorf("expected reboot did not happen")
 }
 
-var shutdownMsg = "reboot scheduled to update the system"
+var (
+	rebootMsg     = "reboot scheduled to update the system"
+	rebootHandler = commandReboot
+)
+
+// SetSyscallReboot replaces the default command-based reboot
+// with a direct Linux kernel syscall based implementation.
+func SetSyscallReboot() {
+	rebootHandler = syscallReboot
+}
 
-func rebootImpl(rebootDelay time.Duration) error {
+// commandReboot assumes a userspace shutdown command exists.
+func commandReboot(rebootDelay time.Duration) error {
 	if rebootDelay < 0 {
 		rebootDelay = 0
 	}
 	mins := int64(rebootDelay / time.Minute)
-	cmd := exec.Command("shutdown", "-r", fmt.Sprintf("+%d", mins), shutdownMsg)
+	cmd := exec.Command("shutdown", "-r", fmt.Sprintf("+%d", mins), rebootMsg)
 	if out, err := cmd.CombinedOutput(); err != nil {
 		return osutil.OutputErr(out, err)
 	}
 	return nil
 }
 
-var reboot = rebootImpl
+var shutdownSyscall = func() {
+	// As per the requirements of the reboot syscall, we have to
+	// first call sync.
+	unix.Sync()
+	// This syscall can fail (EINVAL) if invalid arguments are
+	// supplied, which should not be the case, but let's panic if
```

Comment:
Wouldn't it error if running as a non-root user, too?

Reply:
yes, thanks!

Since the reboot happens non-blocking, I decided to do a pre-check for the required capability (CAP_SYS_BOOT) at the start of the request, before we kick of the timed request. This means EPERM is ruled out, which would make the panic associated with the reboot only possible for an argument error, which means a coding bug, which will be easy to rule out.